### PR TITLE
SAK-49621 Fix dark mode issues in Gateway iframes, input forms, and improve theme scripts

### PR DESCRIPTION
**Pull request by kunaljaykam**

### CODE REVIEW

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 1]
```diff
@@ -49,65 +45,87 @@ portal.getCurrentThemeClass = () => {
...
-    if (portal?.user?.id && portal.userTheme === portal.defaultThemeClass) {
-      themeToApply = portal.isOsDarkThemeSet() ? portal.darkThemeClass : portal.lightThemeClass;
-    } else if (portal.userTheme) {
-      themeToApply = portal.userTheme;
-    }
-    portal.addCssClassToMarkup(themeToApply);
-    portal.updateIframeTheme(themeToApply);
-    portal.addIframeLoadListeners();
+// if the dark theme switch is on the page, attach listener to dark theme toggle switch
+portal.darkThemeSwitcher && portal.darkThemeSwitcher.addEventListener('click', portal.toggleDarkTheme, false);
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:92]:
```suggestion
portal.darkThemeSwitcher?.addEventListener('click', portal.toggleDarkTheme, false);
```

File: library/src/webapp/js/headscripts.js [hunk 2]
```diff
@@ -835,33 +835,27 @@ function includeWebjarLibrary(library, options = {}) {
...
+// switcher script, which applies a user or system-preferred theme class to the document
+if (!window.themeClassInit) {
+	window.themeClassInit = true;
+	document.addEventListener('DOMContentLoaded', () => {
+		if (window.top === window.self && ![...document.documentElement.classList].some(c => c.startsWith('sakaiUserTheme-'))) {
+			const script = document.createElement('script');
+			script.src = '/library/js/portal/portal.theme.switcher.js';
+			script.onload = async () => {
+				try {
+				portal.addCssClassToMarkup(await portal.getCurrentSetTheme());
```
Comment (adrianfish) [inline library/src/webapp/js/headscripts.js:848]:
```suggestion
				    portal.addCssClassToMarkup(await portal.getCurrentSetTheme());
```

File: library/src/webapp/js/headscripts.js [hunk 3]
```diff
@@ -835,33 +835,27 @@ function includeWebjarLibrary(library, options = {}) {
...
+				} catch (error) {
+				console.error('Theme error:', error);
```
Comment (adrianfish) [inline library/src/webapp/js/headscripts.js:850]:
```suggestion
				    console.error('Theme error:', error);
```

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 4]
```diff
@@ -49,65 +45,87 @@ portal.getCurrentThemeClass = () => {
...
+
+  const userId = localStorage.getItem('last-sakai-user');
+  if (!userId) return portal.isOsDarkThemeSet() ? portal.darkThemeClass : portal.lightThemeClass;
+
+  try {
+    const response = await fetch(`/direct/userPrefs/key/${userId}/sakai:portal:theme.json`, {
+      credentials: "include",
+      headers: { 'Cache-Control': 'no-cache', 'Pragma': 'no-cache' }
+    });
+    if (!response.ok) throw new Error('Network error');
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:124]: I would move the url into a const and include it in the error

File: library/src/webapp/js/portal/portal.theme.switcher.js [hunk 5]
```diff
@@ -49,65 +45,87 @@ portal.getCurrentThemeClass = () => {
...
+    const { data } = await response.json();
+    return data?.theme || (portal.isOsDarkThemeSet() ? portal.darkThemeClass : portal.lightThemeClass);
+  } catch (error) {
+    console.error(error.message + ". Using OS theme preference.");
```
Comment (adrianfish) [inline library/src/webapp/js/portal/portal.theme.switcher.js:128]: Use a template literal here

**Review by adrianfish — APPROVED** [review]
Looks ok apart from a few minor things.

**Comment by kunaljaykam** [comment]
Thanks, Adrian!

**Pull request merged**